feat(parental-leave): Edit periods flow

madebynoam opened the pull request:
* Allow a user to edit or add periods after their application is approved.
* Introduce a "Requires action" screen if the application is rejected. To give the user more context on the rejection.

## In Progress
* Adding new tests for the state machine updates

## Screenshots / Gifs

High level overview of updated state chart:
![Parental Leave - Edit Flow](https://user-images.githubusercontent.com/8889977/111760560-c49c9480-8896-11eb-924c-cdee5e378834.png)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/DraftRequiresAction/index.tsx:
```diff
@@ -0,0 +1,98 @@
+  toast.error(
+    formatMessage(
+      {
+        id: 'application.system:submit.error',
```
Would be nice to move it to `libs/application/core/src/lib/messages.ts` and import directly from there

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/DraftRequiresAction/index.tsx:
```diff
@@ -0,0 +1,98 @@
+import { useMutation } from '@apollo/client'
+
+function handleError(error: string, formatMessage: MessageFormatter): void {
+  toast.error(
```
I'm wondering if it would be nice to have an util function for that. I saw it in a couple of places later below, so you won't have to redefine the toast error each time, just a thought

madebynoam replied:
Yeah good catch, refactored it to parentalLeaveUtils.ts

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/DraftRequiresAction/index.tsx:
```diff
@@ -0,0 +1,98 @@
+  )
+}
+
+import { States as ApplicationStates } from '../../lib/ParentalLeaveTemplate'
```
Move the imports with the rest of the imports

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/EditsRequireAction/index.tsx:
```diff
@@ -0,0 +1,126 @@
+import { SUBMIT_APPLICATION } from '@island.is/application/graphql'
+import { useMutation } from '@apollo/client'
+
+function handleError(error: string, formatMessage: MessageFormatter): void {
```
ditto

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/EditsRequireAction/index.tsx:
```diff
@@ -0,0 +1,126 @@
+  )
+}
+
+import { States as ApplicationStates } from '../../lib/ParentalLeaveTemplate'
```
ditto

jeremybarbet commented on libs/application/templates/parental-leave/src/fields/InReviewSteps/index.tsx:
```diff
@@ -1,15 +1,38 @@
+import { getExpectedDateOfBirth } from '../../parentalLeaveUtils'
+import { useMutation } from '@apollo/client'
+
+function handleError(error: string, formatMessage: MessageFormatter): void {
```
ditto

jeremybarbet commented on libs/application/templates/parental-leave/src/forms/EditOrAddPeriods.ts:
```diff
@@ -0,0 +1,117 @@
+  id: 'ParentalLeaveEditOrAddPeriods',
+  title: 'Edit or add periods',
+  logo: Logo,
+  // mode: FormModes.EDITING,
```
missing comment?

madebynoam replied:
oops, brought it back

jeremybarbet commented on libs/application/templates/parental-leave/src/lib/ParentalLeaveTemplate.ts:
```diff
@@ -314,18 +322,235 @@ const ParentalLeaveTemplate: ApplicationTemplate<
+        const { application } = context
+        const { answers } = application
+
+        set(answers, 'periods', cloneDeep(answers['tempPeriods']))
```
```suggestion
        set(answers, 'periods', cloneDeep(answers.tempPeriods))
```

jeremybarbet commented on libs/application/templates/parental-leave/src/lib/ParentalLeaveTemplate.ts:
```diff
@@ -314,18 +322,235 @@ const ParentalLeaveTemplate: ApplicationTemplate<
+        const { application } = context
+        const { answers } = application
+
+        set(answers, 'tempPeriods', answers['periods'])
```
```suggestion
        set(answers, 'tempPeriods', answers.periods)
```

jeremybarbet reviewed (commented):
This is a huge PR, good job on this 🎉  I added a couple of minor comments.

I wonder if it would be nice to get a feature-deploy on it so we can try it out

madebynoam commented:
> This is a huge PR, good job on this 🎉 I added a couple of minor comments.
> 
> I wonder if it would be nice to get a feature-deploy on it so we can try it out

Thanks and thanks for reviewing! Agree it will be nice to try to get this deployed for testing. 
I've added a condition in the state machine (`isDev`), that would allow us to bypass the VMLST states and then we could test the edit flows as well. This will be a great next step.

baering commented on libs/application/templates/parental-leave/src/fields/InReviewSteps/EditsInReviewSteps.tsx:
```diff
@@ -0,0 +1,92 @@
+              parentalLeaveFormMessages.reviewScreen.estimatedBirthDate,
+            )}
+            <br />
+            {format(dobDate, 'dd.MM.yyyy')}
```
The format is stored in a constant [here](https://github.com/island-is/island.is/blob/c0c5e36c3f63d4d0a16421ff55f742541862056c/libs/shared/constants/src/lib/date.ts#L3).

madebynoam replied:
Nice, this should be fixed now.

baering commented on libs/application/templates/parental-leave/src/fields/InReviewSteps/index.tsx:
```diff
@@ -78,16 +94,32 @@ const InReviewSteps: FC<FieldBaseProps> = ({ application }) => {
+              parentalLeaveFormMessages.reviewScreen.estimatedBirthDate,
+            )}
+            <br />
+            {format(dobDate, 'dd.MM.yyyy')}
```
[Same](https://github.com/island-is/island.is/blob/c0c5e36c3f63d4d0a16421ff55f742541862056c/libs/shared/constants/src/lib/date.ts#L3).

madebynoam replied:
Fixed here as well.

baering commented on libs/application/templates/parental-leave/src/forms/EditOrAddPeriods.ts:
```diff
@@ -0,0 +1,115 @@
+                }),
+              ],
+            }),
+            // buildCustomField(
+            //   {
+            //     id: 'endDate',
+            //     name: m.duration,
+            //     description: parentalLeaveFormMessages.duration.description,
+            //     component: 'ParentalLeaveDuration',
+            //   },
+            //   {
+            //     showTimeline: true,
+            //   },
+            // ),
```
Can this be removed?

madebynoam replied:
Yeah I'm not sure why we have this (it is also in ParentalLeaveForm.ts) I've removed it here.

baering commented on libs/application/templates/parental-leave/src/lib/ParentalLeaveTemplate.spec.ts:
```diff
@@ -4,8 +4,10 @@ import {
+  DefaultEvents,
 } from '@island.is/application/core'
 import ParentalLeaveTemplate from './ParentalLeaveTemplate'
+import { States as ApplctionStates } from './ParentalLeaveTemplate'
```
Typo in `ApplctionStates`

baering commented on libs/application/templates/parental-leave/src/lib/ParentalLeaveTemplate.spec.ts:
```diff
@@ -146,11 +148,97 @@ describe('Parental Leave Application Template', () => {
+      const helper = new ApplicationTemplateHelper(
+        buildApplication({
+          answers: {
+            periods,
+            tempPeriods: periods,
+          },
+          state: ApplctionStates.EDIT_OR_ADD_PERIODS,
+        }),
+        ParentalLeaveTemplate,
+      )
+      const [hasChanged, newState, newApplication] = helper.changeState({
+        type: DefaultEvents.ABORT,
+      })
+      expect(hasChanged).toBe(true)
+      expect(newState).toBe(ApplctionStates.APPROVED)
+      expect(newApplication.answers.tempPeriods).toEqual(undefined)
+    })
+
+    it('should assign the application to the employer when the user submits their edits', () => {
+      const helper = new ApplicationTemplateHelper(
+        buildApplication({
+          answers: {
+            employer: {
+              isSelfEmployed: 'no',
+            },
+          },
+          state: ApplctionStates.EDIT_OR_ADD_PERIODS,
+        }),
+        ParentalLeaveTemplate,
+      )
+
+      const [hasChanged, newState, newApplication] = helper.changeState({
+        type: DefaultEvents.SUBMIT,
+      })
+      expect(hasChanged).toBe(true)
+      expect(newState).toBe(
+        ApplctionStates.EMPLOYER_WAITING_TO_ASSIGN_FOR_EDITS,
+      )
+      expect(newApplication.assignees).toEqual([])
+    })
```
Nice 👍

baering reviewed (commented):
Looks good, great work Noam.

I had some issues with reaching the actual edit flow when I tried earlier using the feature deployment, will give it another go soon.

I'll leave these minor comments here for you to look at.

madebynoam commented on libs/application/templates/parental-leave/src/fields/EditPeriodsReview/index.tsx:
```diff
@@ -0,0 +1,56 @@
+  editable?: boolean
+}
+
+const Review: FC<ReviewScreenProps> = ({ application }) => {
```
Ah yes should be fixed now.

madebynoam commented on libs/application/templates/parental-leave/src/fields/EditPeriodsReview/index.tsx:
```diff
@@ -0,0 +1,56 @@
+  const dob = getExpectedDateOfBirth(application)
+
+  if (!dob) {
+    return null
```
Yeah we have this check through out the screens, ideally we should never hit it at this point in the flow. I've refactored it to show a message in case the `dob` is not available.

madebynoam commented on libs/application/templates/parental-leave/src/fields/EditsRequireAction/index.tsx:
```diff
@@ -0,0 +1,109 @@
+                if (res?.data) {
+                  // Takes them to the next state (which loads the relevant form)
+                  refetch?.()
+                }
```
Good catch, fixed.

madebynoam commented:
> Looks good, great work Noam.
> 
> I had some issues with reaching the actual edit flow when I tried earlier using the feature deployment, will give it another go soon.
> 
> I'll leave these minor comments here for you to look at.

Awesome thanks for the review! Should be in good shape now.

jeremybarbet reviewed (approved):
LGTM, been able to try out the flow and add a new period it was working well! 👍  I believe there are some design adjustments to do here and there but we might want to wait on the application system design system to go over that on another PR

madebynoam commented:
> LGTM, been able to try out the flow and add a new period it was working well! 👍 I believe there are some design adjustments to do here and there but we might want to wait on the application system design system to go over that on another PR

Nice, thanks for the review! Happy to hear you were able to try it out.
Agree, will be good to do a pass on clean up and polish when the design system is ready.
